src/Titanium.Web.Proxy/Certificates/CertificateManager.cs

```diff
@@ -503,33 +507,50 @@ private X509Certificate2 makeCertificate(string certificateName, bool isRootCert
                 return cached.Certificate;
             }
 
-            // handle burst requests with same certificate name
-            // by checking for existing task for same certificate name.
-            // If two or more requests hit this block at the same time, then multiple tasks will be created,
-            // which is okay. That certificate will be created twice or more. We don't anticipate many requests for same host at the same time.
-            // This saves us from another expensive lock. The goal here is to minimize creation of multiple tasks for same certification name.
-            // Goal is not to guarantee single certificate creation, which is not needed in our case.
-            if (pendingCertificateCreationTasks.TryGetValue(certificateName, out var task))
-            {
-                return await task;
-            }
-
-            // run certificate creation task & add it to pending tasks         
-            task = Task.Run(() =>
-            {
-                var result = CreateCertificate(certificateName, false);
-                if (result != null)
+            var createdTask = false;
+            Task<X509Certificate2?> createCertificateTask;
+            await pendingCertificateCreationTaskLock.WaitAsync();
```

Comment:
Since this critical section code would be very fast and we don't anticipate any delay inside the section, a synchronous lock like [SpinLock](https://docs.microsoft.com/en-us/dotnet/api/system.threading.spinlock?view=net-5.0) may perform better than an async await lock. But good for now, we can try that in future. You can test with a sync lock and see if it perform better.

Reply:
Task creation can probably be slow if there are too many tasks running.    Could be worth testing though.